### DISCUSSION TITLE
Botan::TLS::Server, TLS::Client VS. boost::asio, boost::beast

### DISCUSSION THREAD

**Discussion by fasxmut** [randombit/botan#2994] [Q&A] [answered]
What is the difference between Botan::TLS::Server, TLS::Client and boost::asio, boost::beast? Does botan implement its own network api?

**Comment by randombit** [comment, marked as answer]
I'm not familiar with the Beast library so cannot comment there.

With regards to TLS Botan does not implement the network layer. The `TLS::Server` and `TLS::Client` classes provide a wrapper around some other network API which you as the application developer must provide. This is done by implementing various functions in `TLS::Callbacks` for example `tls_emit_data` is called when the TLS layer needs to send something to the network. At that point it is up to the application and it might well use something like asio to accomplish that.

It may clear things up to look at some of the command line utility implementations like [tls_http_server](https://github.com/randombit/botan/blob/master/src/cli/tls_http_server.cpp) which use asio to access the network layer.

**Reply by reneme** [reply]
It may be worth mentioning that Botan also provides an adapter for Boost asio that aims to be compatible with Asio's ssl::stream. With that it is easy to integrate Botan's TLS implementation with networking provided by asio.

See here for an example HTTPS client based on asio, beast and Botan: https://github.com/hrantzsch/beast-botan-https-client

**Reply by reneme** [reply]
For what it's worth: There's also a blog post on how to use Botan to provide TLS for Boost ASIO and beast and essentially write a minimal HTTPS client: https://botan-crypto.org/using-botan-tls-with-boost-beast/

**Comment by fasxmut** [comment]
Thank you for the answer.
